### PR TITLE
Review request for re-former project

### PR DESCRIPTION
**Pull request by chrystalme**
# Re-former project

- To create a form in rails using form_with, form_tag, form_for
- Using form partials

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+  ``` rails 6.1.7 ```
+  ``` yarn 1.2.2 ```
+  ``` node 14.15.5 ```
+
+### Configuration
+
+  Set up project using the above specifications
+   #### For HTTPS
+   - use ``` https://github.com/chrystalme/re-former-project.git ``` to clone the repository
```
Comment (whiz25) [inline README.md:24]: - Add explicit command for cloning: `git clone ...`
- How does one start the project after cloning? Add instructions for changing to the `feature/re-former` branch, installing `yarn`, install packages using `yarn install`, bundling gems in rails, starting the rails server and migrating the database

File: README.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+   #### For SSH
+   - use ``` git@github.com:chrystalme/re-former-project.git ``` to clone the repository
+   - Change to the parent directory of the project using 
+    ``` cd re-former-project ```
+   - run ```bundle install ``` to install dependencies
```
Comment (whiz25) [inline README.md:29]: - Add instruction for checking out the `feature/re-former` branch before install gems.

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
+class UsersController < ApplicationController
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(username: params[:username], email: params[:email], password: params[:password])
```
Comment (whiz25) [inline app/controllers/users_controller.rb]: - This is redundant. You already have logic for this in `user_params` method
```ruby
@user = User.new(user_params)
```

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+  def edit
+    @user = User.find(params[:id])
+  end
+
+  def update
+    @user = User.find(params[:id])
+
+    if @user.update(user_params)
+      redirect_to @user
```
Comment (whiz25) [inline app/controllers/users_controller.rb]: - Change this to `redirect_to edit_user_path`

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    else
+      render :edit, status: :unprocessable_entity
+    end
+  end
+
+  private
+
+  # def set_user
+  #   @user = User.find(params[:id])
+  # end
```
Comment (whiz25) [inline app/controllers/users_controller.rb]: - Remove this comment. I makes your codebase look ugly.

File: app/views/users/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+<%# <form method='POST' action='/users' accept-charset="UTF-8"> %>
+<%# <input type="hidden" name="authenticity_token" value="<%= form_authenticity_token %>
+<%# <label for='usename'>Username: </label> %>
+<%# <input type='text' name='user[username]' placeholder='Enter your username'><br> %>
+<%# <label for='email'>Email: </label> %>
+<%# <input type='email' name='user[email]' placeholder='Enter your email'><br> %>
+<%# <label for='password'>Password: </label> %>
+<%# <input type='password' name='user[password]' placeholder='Enter password at least 8 alphanum'><br> %>
+<%# <input value='submit' type='submit'> %>
+<%# </form> %>
```
Comment (whiz25) [inline app/views/users/new.html.erb]: - This form is not properly commented out. A simple trick is to highlight the code then press and hold `CTRL+/`
Below is a properly commented out code:
```ruby
# <%= form_with model: @user do |form| %>
#     <%= label_tag(:username, "Username:") %><br>
#     <%= form.text_field :username %><br><br>
#     <%= label_tag(:email, "email:") %><br>
#     <%= form.email_field :email %><br><br>
#     <%= label_tag(:password, "Password:") %><br>
#     <%= form.password_field :password %><br><br>
#     <%= form.submit :Submit %>
# <% end %>
```

File: app/views/users/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+	 	<%# <%= label_tag(:email, "Email") %>
+		<%# <%= email_field_tag(:email) %> 
+	<%# </p> %>
+
+	<%# <p> %>
+		<%# <%= label_tag(:password, "Password") %> 
+  	<%# <%= password_field_tag(:password) %> 
+	<%# </p> %>
+	  <%# <%= submit_tag("Submit") %> 
+<%# <% end %>
```
Comment (whiz25) [inline app/views/users/new.html.erb]: - Properly comment out this code. See suggestion above.

File: app/views/users/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+	<%# <p> %>
+		<%# <%= label_tag(:password, "Password") %> 
+  	<%# <%= password_field_tag(:password) %> 
+	<%# </p> %>
+	  <%# <%= submit_tag("Submit") %> 
+<%# <% end %> 
+
+
+<%= render 'form', user: @user %>
+<a href="/">Go Back</a>
```
Comment (whiz25) [inline app/views/users/new.html.erb]: - You are lacking the form for actually creating the user using `form_for`
Please add a new form for creating a user:
```ruby
<%= form_with @user do |u|%> <%= u.label :username %> <%= u.text_field :username%> 
<%= u.label :email %> 
<%= u.email_field :email %> 
<%= u.label :password %>
<%= u.password_field :password %> 
<%= u.submit %> <% end %>
```

File: app/views/users/_form.html.erb [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+
+  <%= form.email_field :email, placeholder: "Email@example.com" %><br />
+  <%= user.errors.full_messages_for(:email_address).first %><br /> 
+
+  <%= form.password_field :password, placeholder: "Password" %><br />
+  <%= user.errors.full_messages_for(:password).first %><br /> 
+
+  <%= form.submit %>
+<% end %>
```
Comment (whiz25) [inline app/views/users/_form.html.erb]: - Move the code form here to `new.html`

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Changes Required :pray:
Nice work on this PR. You managed to set the project app and modeled the user data properly. Well done!
Lets however, make some changes before we can merge.

 - [ ] Right now the app is unable to create a new user and update the user created. Check my comments form suggestions.

### Next Projects
Follow the following guidelines in your next projects.
- [ ] Be consistent in your commit messages: `add 'forM_with' '` and `Update README.md` are quite different in format. The accepted format is to capitalize the first letter of the commit message like this: `Add 'form_with'`

### Martin Afani

File: app/views/users/new.html.erb [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  <%= form.text_field :username, placeholder: "Username" %><br />
+
+  <%= form.email_field :email, placeholder: "Email@example.com" %><br />
+
+  <%= form.password_field :password, placeholder: "Password" %><br />
+
+	<%= form.submit %>
+<% end %> --->
+
+<%= form_for(@user) do |form|%>
```
Comment (SmithVyne) [inline app/views/users/new.html.erb]: Remove the `form_for` tag and render your `form` partial here instead. Hint:
`<%= render 'form'%>`

**Review by SmithVyne — CHANGES_REQUESTED** [review]
### Changes Required 👍 🤕 
Hello, congratulations on making the changes requested by the previous TSE but you need to make a few more improvements to this project to this project to get it approved.
[REQUIRED] - Use the `users_path` route available in Rails instead of `/users` in all your forms.
[REQUIRED] - You should be using the `form_with` tag instead of the `form_for` tag in the new users view.

You can reach out to me on slack, @smithvyne if you have any more questions.

**Review by whiz25 — APPROVED** [review]
## Approved ✅ 
Well done working on the requested changes. A user can now signup and edit their details. Merge and proceed to the next project.


### Martin Afani

**Pull request merged**